### ISSUE TITLE
Running out of connections

### ISSUE THREAD

**Issue by INRIX-Cory-Bradshaw** [mrniko/redisson#386]
I use Elasticache Redis as my redis server, 1 master, 1 slave (fairly large machine, this is an autoscaling micro webservice).  I used to use an older version and it worked good (from a connections point of view), but I had to upgrade from 2.1.2 due to distributed lock fixes. I am currently using 2.2.4, but after about 12-24 hours I see the " RedisConnectionException:Can't aquire connection from pool! " exceptions. 

I expected it had to do with Elasticache closing idle connections, so I turned that back off for now, but it didn't effect this behavior, it seems to be related with my update from 2.1.2 to 2.2.4.

Does redisson not refresh its connection pool as connections close, do they timeout, or is this something else entirely?

I am using SingleServer or MasterSlave Configuration, with defaults, the current behavior is seen with SingleServer. (I haven't added configurations for slaves yet)

The only functionality I currently use in Redisson is the Distributed lock mechanism, to control 2 sets of locks.

Flow of using lock: get lock, set expire to 15s, (try) process,(finally) unlock (or set lock to expire in 500ms)

Redis has plenty of connections available. only 1300 or so used out of about 7k

**Comment by mrniko** [comment]
Redisson has async connection pool since 2.1.4 version, so connection acquisition approach has changed entirely.

When `Can't aquire connection from pool!` occurs it has an additional info like `entires` field state. Could you please print it too.

**Comment by INRIX-Cory-Bradshaw** [comment]
Extra Details as requested: (address hidden)
	
org.redisson.client.RedisConnectionException:Can't aquire connection from pool! [[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=0, freeConnectionsAmount=0, freeConnectionsCounter=0, freezed=false, freezeReason=null, client=[addr=fubar.com/000.00.00.00:6379], nodeType=MASTER, failedAttempts=0]]\nat org.redisson.misc.ConnectionPool.get(ConnectionPool.java:124)at org.redisson.connection.MasterSlaveEntry.connectionWriteOp(MasterSlaveEntry.java:184)at org.redisson.connection.MasterSlaveConnectionManager.connectionWriteOp(MasterSlaveConnectionManager.java:565)at org.redisson.command.CommandAsyncService.async(CommandAsyncService.java:341)at org.redisson.command.CommandAsyncService$5.run(CommandAsyncService.java:386)at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:655)at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)at java.lang.Thread.run(Thread.java:745)

**Comment by mrniko** [comment]
> Does redisson not refresh its connection pool as connections close, do they timeout, or is this something else entirely?

It has ConnectionWatchdog it handles situation when connection is down and tries to reconnect.

**Comment by mrniko** [comment]
do you see in logs messages produced by this line at ConnectionWatchdog:100?
`log.warn("Can't connect " + connection + " to " + connection.getRedisClient().getAddr(), e);`

**Comment by INRIX-Cory-Bradshaw** [comment]
No, no messages containing "Can't connect" in the logs with severity of Warning. We use logback appenders (Dropwizard framework) if that matters any.

**Comment by mrniko** [comment]
Could you try to lunch it with debug level mode for `org.redisson` package ?

**Comment by INRIX-Cory-Bradshaw** [comment]
I'll try it and hope it works. Dropwizard tends to override most configurations due to their glueing tools together.

**Comment by mrniko** [comment]
Please note debug lines like at CommandAsyncService:541: 
```java
            log.debug("connection released for command {} and params {} from slot {} using node {}",
                    details.getCommand(), Arrays.toString(details.getParams()), details.getSource(), connection.getRedisClient().getAddr());
```

**Comment by INRIX-Cory-Bradshaw** [comment]
It will be deployed in a few days and I'll respond to this ticket with more details as I find them. Thanks for your help!

**Comment by mrniko** [comment]
try 2.2.7 version which has just gone to maven central and increase connection pool size amount.

**Comment by mrniko** [comment]
@INRIX-Cory-Bradshaw Have you tried?

**Comment by INRIX-Cory-Bradshaw** [comment]
I couldn't update to latest due to dependency conflicts with the web service framework we are using. Thanks for your help though

**Comment by codenative** [comment]
Just wanted to add that I am also getting the "Can't aquire connection from pool!" error when running a load test with both Redisson 2.2.7 and 2.2.8 (error message is different but is about the exhausted pool).
I downgraded Redisson to 2.1.2 and the same load test succeeds without the error.
As an idea of the load I am running: 
- v2.1.2 opens 38 of the 100 max connection in version 2.1.2 and happily continues.
- v2.2.8 uses the entire 100 connections and falls over exhausted pool error.
- I have also tried increased the max connections to 1000 in v2.2.8. Redisson quickly uses all 1k connections and throws the exhausted error message again.

**Comment by INRIX-Cory-Bradshaw** [comment]
Thanks for adding more information. Is there any way we can get the connection handling of 2.1.2 (perhaps a configuration setting?)  with the fixes in 2.2.4 and on? My biggest concern at the moment is that I will be unable to use any future fixes due to being locked in on a lower version of Jackson that is incompatible with the newest releases.

Was the upgrade in Jackson necessary, or just to keep with the latest?

**Comment by mrniko** [comment]
Have you tried latest Redisson with your Jackson version? I think it should work.

**Comment by mrniko** [comment]
Have you tried latest Redisson with your Jackson version? I think it should work.

**Comment by INRIX-Cory-Bradshaw** [comment]
I tried, but it failed, let me try again to make sure i have the exclusions in the right places.

**Comment by mrniko** [comment]
Yeah, try again please. And check you classpath for Jackson libs version mix, 2.4.x with 2.5.x, for example.

**Comment by INRIX-Cory-Bradshaw** [comment]
The application deployment is using shade, so it can't have multiple versions for the same library.

**Comment by INRIX-Cory-Bradshaw** [comment]
OK, got the exclusions setup correctly. (I was missing an exclusions for the yaml artifact) How can we address the connection problem @codenative mentioned?

**Comment by mrniko** [comment]
@codenative which command execution mode do you use, asynchronous?

**Comment by codenative** [comment]
Synchronous

**Comment by mrniko** [comment]
could you share your testcase?

**Comment by codenative** [comment]
I can reproduce the issue in 2.2.2 but it is working fine in 2.2.1 so it may be caused by changes in version 2.2.2.

**Comment by mrniko** [comment]
I don't see in git history any changes of command execution logic between 2.2.1 and 2.2.2 versions

**Comment by mrniko** [comment]
I have some ideas how connection pool could be improved. Some thinking is needed

**Comment by codenative** [comment]
@mrniko Okay. Thanks for your work btw, great client!

**Comment by INRIX-Cory-Bradshaw** [comment]
@mrniko, I second @codenative comment,  thanks for the wonderful support.

**Comment by mrniko** [comment]
Thanks for you using and patience, guys!

**Comment by mrniko** [comment]
I created this task for connection pool enhancements https://github.com/mrniko/redisson/issues/414. Please track it

**Comment by mrniko** [comment]
@codenative hi! Could you please try to run your tests with version from `cpv2` branch?

**Comment by codenative** [comment]
@mrniko I just tested with the same load as last time and it's working nicely without running out of connections. Great work!

**Comment by mrniko** [comment]
@codenative there is a commands queue with size 3 elements for each channel. I'm just investigating which is better with queue or without it.

**Comment by codenative** [comment]
@mrniko I just tried to reproduce it with v2.2.8 and v2.2.9 and it's not having the same connection issue so I'm a bit stumped. We were able to reproduce the issue in multiple environments previously so I am not sure as to why it isn't doing it anymore. It could be an intermittent issue so I will keep an eye on it.

**Comment by mrniko** [comment]
@codenative 

The issue was introduced in 2.3.0 version and fixed in https://github.com/redisson/redisson/issues/2587
